libraries/bot-ai-qna/src/main/java/com/microsoft/bot/ai/qna/QnAMaker.java

```diff
@@ -0,0 +1,338 @@
+        return fillQnAEvent(queryResults, turnContext, telemetryProperties, telemetryMetrics).thenAccept(eventData -> {
+            this.telemetryClient.trackEvent(QnATelemetryConstants.QNA_MSG_EVENT,
+                eventData.getLeft(), eventData.getRight());
+        });
+    }
+
+    /**
+     * Fills the event properties and metrics for the QnaMessage event for telemetry.
+     * These properties are logged when the QnA GetAnswers method is called.
+     * @param queryResults QnA service results.
+     * @param turnContext Context object containing information for a single turn of conversation with a user.
+     * @param telemetryProperties Properties to add/override for the event.
+     * @param telemetryMetrics Metrics to add/override for the event.
+     * @return A tuple of Properties and Metrics that will be sent to the IBotTelemetryClient.
+     * TrackEvent method for the QnAMessage event.
+     * The properties and metrics returned the standard properties logged with any properties passed from the
+     * GetAnswersAsync method.
+     */
+    protected CompletableFuture<Pair<Map<String, String>, Map<String, Double>>> fillQnAEvent(QueryResult[] queryResults,
+                                                                                             TurnContext turnContext,
+                                                                                             @Nullable Map<String, String> telemetryProperties,
+                                                                                             @Nullable Map<String, Double> telemetryMetrics) {
+        Map<String, String> properties = new HashMap<>();
+        Map<String, Double> metrics = new HashMap<>();
+
+        properties.put(QnATelemetryConstants.KNOWLEDGE_BASE_ID_PROPERTY, endpoint.getKnowledgeBaseId());
+
+        String text = turnContext.getActivity().getText();
+        String userName = turnContext.getActivity().getFrom().getName();
+
+        // Use the LogPersonalInformation flag to toggle logging PII data, text and user name are common examples
+        if (this.logPersonalInformation) {
+            if (!StringUtils.isWhitespace(text)) {
+                properties.put(QnATelemetryConstants.QUESTION_ID_PROPERTY, text);
+            }
+
+            if (!StringUtils.isWhitespace(userName)) {
+                properties.put(QnATelemetryConstants.USERNAME_PROPERTY, userName);
+            }
+        }
+
+        // Fill in QnA Results (found or not)
+        if (queryResults.length > 0) {
+            JacksonAdapter jacksonAdapter = new JacksonAdapter();
+            QueryResult queryResult = queryResults[0];
+            try {
+                properties.put(QnATelemetryConstants.MATCHED_QUESTION_PROPERTY,
+                    jacksonAdapter.serialize(queryResult.getQuestion()));
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
```

Comment:
I'd prefer to have the signatura throwable IOException instead of this handling. See another examples